doc/cpp_hardware.rst

```diff
@@ -0,0 +1,392 @@
+separation of concerns, so that users have full control over how hardware resources
+are selected, either using the functions and types in *libarborenv*, or writing their
+own code for managing MPI, GPUs, and thread counts.
+
+.. cpp:namespace:: arbenv
+
+.. cpp:function:: arb::util::optional<int> get_env_num_threads()
+
+    Tests whether the number of threads to use has been set in an environment variable.
+    First checks ``ARB_NUM_THREADS``, and if that is not set checks ``OMP_NUM_THREADS``.
+
+    Return value:
+
+    * no value: the :cpp:any:`optional` return value contains no value if the
+    * has value: the number of threads set by the environment variable.
+
+    Throws:
+
+    * throws :cpp:any:`std::runtime_error` if environment variable set with invalid
+      number of threads.
+
+    .. container:: example-code
+
+       .. code-block:: cpp
+
+         #include <arborenv/concurrency.hpp>
+
+         if (auto nt = arbenv::get_env_num_threads()) {
+            std::cout << "requested " << nt.value() << "threads \n";
+         }
+         else {
+            std::cout << "no enviroment variable set\n";
+         }
+
+.. cpp:function:: int thread_concurrency()
+
+   Attempts to detect the number of available CPU cores. Returns 1 if unable to detect
+   the number of cores.
+
+    .. container:: example-code
+
+       .. code-block:: cpp
+
+         #include <arborenv/concurrency.hpp>
+
+         // Set num_threads to value from environment variable if set,
+         // otherwise set it to the available number of cores.
+         int num_threads = 0;
+         if (auto nt = arbenv::get_env_num_threads()) {
+            num_threads = nt.value();
+         }
+         else {
+            num_threads = arbenv::thread_concurrency();
+         }
+
+.. cpp:function:: int default_gpu()
+
+   Returns the integer identifier of the first available GPU, if a GPU is available 
+
+   Return value:
+
+   * non-negative value: if a GPU is available, the index of the selected GPU is returned. The index will be in the range ``[0, num_gpus)`` where ``num_gpus`` is the number of GPUs detected using the ``cudaGetDeviceCount`` `CUDA API call <https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__DEVICE.html>`_.
+   * -1: if no GPU available, or if Arbor was built without GPU support.
+
+    .. container:: example-code
+
+       .. code-block:: cpp
+
+         #include <arborenv/gpu_env.hpp>
+
+         if (arbenv::default_gpu()>-1) {}
+            std::cout << "a GPU is available\n";
+         }
+
+.. cpp:function:: int find_private_gpu(MPI_Comm comm)
+
+   A helper function that assigns a unique GPU to every MPI rank.
+
+   .. Note::
+
+      Arbor allows at most one GPU per MPI rank, and furthermore requires that
+      an MPI rank has exclusive access to a GPU, i.e. two MPI ranks can not
+      share a GPU.
+      The task of assigning a unique GPU to each rank when more than one rank
+      have access to the same GPU(s).
```

Comment:
The sentence sounds a bit odd; maybe use "This function assigns a unique GPU to each rank when more than one rank has access to the same GPU(s)"